images/virtualization-artifact/pkg/controller/vd/internal/source/upload.go

```diff
@@ -0,0 +1,388 @@
+	"github.com/deckhouse/virtualization/api/core/v1alpha2/vdcondition"
+)
+
+type UploadDataSource struct {
+	statService     *service.StatService
+	uploaderService *service.UploaderService
+	diskService     *service.DiskService
+	dvcrSettings    *dvcr.Settings
+	logger          *slog.Logger
+}
+
+func NewUploadDataSource(
+	statService *service.StatService,
+	uploaderService *service.UploaderService,
+	diskService *service.DiskService,
+	dvcrSettings *dvcr.Settings,
+) *UploadDataSource {
+	return &UploadDataSource{
+		statService:     statService,
+		uploaderService: uploaderService,
+		diskService:     diskService,
+		dvcrSettings:    dvcrSettings,
+		logger:          slog.Default().With("controller", common.VDShortName, "ds", "upload"),
+	}
+}
+
+func (ds UploadDataSource) Sync(ctx context.Context, vd *virtv2.VirtualDisk) (bool, error) {
+	ds.logger.Info("Sync", "vd", vd.Name)
+
+	condition, _ := service.GetCondition(vdcondition.ReadyType, vd.Status.Conditions)
+	defer func() { service.SetCondition(condition, &vd.Status.Conditions) }()
+
+	supgen := supplements.NewGenerator(common.VDShortName, vd.Name, vd.Namespace, vd.UID)
+	pod, err := ds.uploaderService.GetPod(ctx, supgen)
+	if err != nil {
+		return false, err
+	}
+	svc, err := ds.uploaderService.GetService(ctx, supgen)
+	if err != nil {
+		return false, err
+	}
+	ing, err := ds.uploaderService.GetIngress(ctx, supgen)
+	if err != nil {
+		return false, err
+	}
+	dv, err := ds.diskService.GetDataVolume(ctx, supgen)
+	if err != nil {
+		return false, err
+	}
+	pvc, err := ds.diskService.GetPersistentVolumeClaim(ctx, supgen)
+	if err != nil {
+		return false, err
+	}
+	pv, err := ds.diskService.GetPersistentVolume(ctx, pvc)
+	if err != nil {
+		return false, err
+	}
+
+	if vd.Status.UploadCommand == "" {
+		if ing != nil && ing.Annotations[common.AnnUploadURL] != "" {
+			vd.Status.UploadCommand = fmt.Sprintf("curl %s -T example.iso", ing.Annotations[common.AnnUploadURL])
+		}
+	}
+
+	switch {
+	case isDiskProvisioningFinished(condition):
+		ds.logger.Info("Finishing...", "vd", vd.Name)
+
+		switch {
+		case pvc == nil:
+			condition.Status = metav1.ConditionFalse
+			condition.Reason = vdcondition.ReadyReason_Lost
+			condition.Message = fmt.Sprintf("PVC %s not found.", supgen.PersistentVolumeClaim().String())
+		case pv == nil:
+			condition.Status = metav1.ConditionFalse
+			condition.Reason = vdcondition.ReadyReason_Lost
+			condition.Message = fmt.Sprintf("PV %s not found.", pvc.Spec.VolumeName)
+		default:
+			condition.Status = metav1.ConditionTrue
+			condition.Reason = vdcondition.ReadyReason_Ready
+			condition.Message = ""
+		}
+
+		err = ds.diskService.Protect(ctx, vd, nil, pvc, pv)
+		if err != nil {
+			return false, err
+		}
+
+		err = ds.uploaderService.Unprotect(ctx, pod, svc, ing)
+		if err != nil {
+			return false, err
+		}
+
+		err = ds.diskService.Unprotect(ctx, dv)
+		if err != nil {
+			return false, err
+		}
```

Comment:
Am I correct about these protect/unprotect session?

```suggestion
		// Protect Ready Disk and underlying PVC and PV.
		err = ds.diskService.Protect(ctx, vd, nil, pvc, pv)
		if err != nil {
			return false, err
		}

		// Unprotect upload time supplements to delete them later.
		err = ds.uploaderService.Unprotect(ctx, pod, svc, ing)
		if err != nil {
			return false, err
		}

		err = ds.diskService.Unprotect(ctx, dv)
		if err != nil {
			return false, err
		}
```

Reply:
Yes, added these comments